File: TreeFromPreorder.py
# Problem Description: https://practice.geeksforgeeks.org/problems/construct-tree-from-preorder-traversal/1

class Node:
    def __init__(self, val):
        self.data = val
        self.right = None
        self.left = None


def printInorder(root):
    if(not root):
        return
    if(root.left):
        printInorder(root.left)
    print(root.data, end=" ")
    if(root.right):
        printInorder(root.right)


def constructTree(n, pre, preLN):
    root = Node(pre[0])
    if(n == 1):
        return root

    q = [root]
    i = 1
    temp = root
    while(i < n):
        curr_node = Node(pre[i])
        if(preLN[i] == 'N'):
            if(not temp.left):
                temp.left = curr_node
                temp = temp.left
            else:
                temp = q.pop()
                temp.right = curr_node
                temp = curr_node
            q.append(curr_node)
        elif(preLN[i] == 'L' and preLN[i-1] == 'N'):
            temp.left = curr_node
        elif(preLN[i] == 'L'):
            temp = q.pop()
            temp.right = curr_node
        i += 1

    return root


if __name__ == "__main__":
    t = int(input())

    for _ in range(t):
        n = int(input())
        pre = [int(x) for x in input().split()]
        preLN = [x for x in input().split()]

        root = constructTree(n, pre, preLN)
        printInorder(root)
        print()

File: TreeFromPreorder.cpp
// Problem Description: https://practice.geeksforgeeks.org/problems/construct-tree-from-preorder-traversal/1

#include <bits/stdc++.h>

using namespace std;

struct Node
{
    int data;
    struct Node *left;
    struct Node *right;
};
struct Node *newNode(int data)
{
    Node *temp = new Node;
    temp->data = data;
    temp->left = NULL;
    temp->right = NULL;
    return temp;
}
void printInorder(struct Node *node)
{
    if (node == NULL)
        return;
    printInorder(node->left);
    cout << node->data << " ";
    printInorder(node->right);
}

struct Node *constructTree(int n, int pre[], char preLN[])
{
    Node *root = newNode(pre[0]);
    if (n == 1)
        return (NULL);

    stack<Node *> st;
    st.push(root);
    int i = 1;
    Node *temp = root;
    while (i < n)
    {
        Node *curr_node = newNode(pre[i]);
        if (preLN[i] == 'N')
        {
            if (temp->left == NULL)
            {
                temp->left = curr_node;
                temp = temp->left;
            }
            else
            {
                temp = st.top();
                st.pop();
                temp->right = curr_node;
                temp = curr_node;
            }
            st.push(curr_node);
        }
        else if (preLN[i] == 'L' && preLN[i - 1] == 'N')
        {
            temp->left = curr_node;
        }
        else if (preLN[i] == 'L')
        {
            temp = st.top();
            st.pop();
            temp->right = curr_node;
        }
        i++;
    }
    return (root);
}

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        Node *root = NULL;
        int n;
        cin >> n;
        int pre[n];
        char preLN[n];
        for (int i = 0; i < n; i++)
            cin >> pre[i];
        for (int i = 0; i < n; i++)
            cin >> preLN[i];
        root = constructTree(n, pre, preLN);
        printInorder(root);
        cout << endl;
    }
    return 0;
}